lib/datatype/index.js

```diff
@@ -108,4 +155,164 @@ export class DataType {
+      return this.#writer.info()
+    }
+
+    return Promise.all(
+      this.cores.map(async (core) => {
+        return core.info()
+      })
+    )
+  }
+
+  /**
+   * Get a doc by id
+   * @param {string} id
+   * @returns {Doc}
+   */
+  getById(id) {
+    return this.#sqlite.get(`select * from ${this.name} where id = '${id}'`)
+  }
+
+  /**
+   * Put a doc
+   * @param {Doc} data
+   * @returns {Promise<Doc>}
+   */
+  async put(data) {
+    await this.#writer.ready()
+
+    const created = data.created || new Date().getTime()
+    const doc = Object.assign({}, data, {
+      id: data.id || randomBytes(8).toString('hex'),
+      authorId: this.identityId,
+      version: `${this.#writer.key.toString('hex')}@${this.#writer.length}`,
+      created,
+      timestamp: created,
+    })
+
+    if (!doc.links) {
+      doc.links = []
+    }
+
+    if (!doc.links.length) {
+      const existing = this.#sqlite.query(
+        `select version from ${this.name} where id = '${doc.id}' ORDER BY timestamp`
+      )
+
+      if (existing) {
+        /**
+         * @type {string[]}
+         */
+        doc.links = existing.map((row) => {
+          return row.version
+        })
+      }
+    }
+
+    const encodedDoc = this.encode(doc)
+    const indexing = new Promise((resolve) => {
+      this.#indexer.onceWriteDoc(doc.version, (doc) => {
+        resolve(doc)
+      })
+    })
+
+    await this.#writer.append(encodedDoc)
+    await indexing
+    return doc
+  }
+
+  /**
+   * Create a doc
+   * @param {Doc} data
+   * @returns {Promise<Doc>}
+   */
+  async create(data) {
+    return this.put(data)
+  }
+
+  /**
+   * Update a doc
+   * @param {Doc} data
+   * @returns {Promise<Doc>}
+   */
+  async update(data) {
+    const doc = Object.assign({}, data, {
+      updated: new Date().getTime(),
+    })
+
+    return this.put(doc)
+  }
+
+  /**
+   * Check if a block is of this DataType
+   * @param {Block} block
+   * @returns {boolean}
+   */
+  isType(block) {
+    return getBlockPrefix(block) === this.blockPrefix
+  }
+
+  /**
+   * Index an array of blocks
+   * @param {Block[]} blocks
+   * @returns {void}
+   */
+  index(blocks) {
+    const docs = blocks.map((block) => {
+      return this.decode(block)
+    })
+
+    this.#indexer.batch(docs)
+  }
+
+  /**
+   * Query indexed docs
+   * @param {string} sql sql statement
+   * @param {any[]} params
+   * @returns {Doc[]}
+   */
+  query(sql, params) {
+    return this.#sqlite.query(sql, params)
+  }
+
+  /**
+   * Create a read stream of blocks from the local writer hypercore
+   * @param {Object} [options]
+   * @return {import('streamx').Readable}
+   */
+  createReadStream(options) {
+    return this.#writer.createReadStream(options)
+  }
+
+  /**
+   * Create a read stream of blocks from the local writer hypercore
+   * @return {Promise<void>}
+   */
+  async close() {
+    await this.#writer.close()
+    await this.#corestore.close()
+    await this.#sqlite.close()
```

Comment:
Promise.all()? or is order important?

Reply:
Also, I think a class should only be responsible for closing instances that it "owns". E.g. if this class did not "open" these instances, then it should not close them.